File: src/FlowField.java
import processing.core.PApplet;
import processing.core.PVector;
@SuppressWarnings("unchecked")
public class FlowField extends PApplet {

    public static void main (String... args) {
        FlowField pt = new FlowField();
        PApplet.runSketch(new String[]{"FlowField"}, pt);
    }

    int scl = 20;
    int refreshRate = 1000;

    float fieldStrength = (float) 0.3;

    float zoffInc = (float) 0.01;
    int angleMult = 2;

    float zoff = 0;
    float noiseInc = (float) 0.1;

    float decline = (float) 0.9999;

    int rows, cols;

    PVector[] flowfield;
    Particle[] particles = new Particle[10000];

    float h = 0;
    boolean colUp = true;

    @Override
    public void settings() {
        size(1920, 1080, P2D);
        fullScreen(P2D);

        cols = width/scl;
        rows = height/scl;

        flowfield = new PVector[rows*cols];

        createParticles();


    }

    @Override
    public void setup() {
        colorMode(HSB);
        background(0);
    }

    @Override
    public void draw() {
            loadPixels();
            for(int i = 0; i < pixels.length; ++i){
                int rgb = pixels[i];
                int temp = (int) (brightness(rgb) * decline);
                pixels[i] = color(hue(rgb),saturation(rgb),temp);

            }
            updatePixels();


        float yoff = 0;
        for (int y = 0; y < rows; y++) {
            float xoff = 0;
            for (int x = 0; x < cols; x++) {
                int index = x+y*cols;
                float a = noise(xoff, yoff, zoff) * TWO_PI*angleMult;
                xoff += noiseInc;

                PVector v = PVector.fromAngle(0);
                v.rotate(a);
                v.setMag(fieldStrength);

                flowfield[index] = v;
            }
            yoff += noiseInc;
        }
        zoff += zoffInc;

        for (int i = 0; i < particles.length; i++) {
            particles[i].updateColor(color(h, 255, 255));
            particles[i].edges();
            particles[i].follow(flowfield);
            particles[i].update();
            particles[i].display();
        }


        if (colUp) {
            h++;
        } else {
            h--;
        }

        if (h > 255) {
            colUp =false;
            h = 255;
        } else if(h < 0) {
            colUp = true;
            h = 0;
        }

    }

    @Override
    public void mouseClicked() {
        if (mouseX < width / 2) {
            background(0);
        }else{
            refresh();
        }
    }

    void refresh() {
        background(0);
        createParticles();
    }

    void createParticles() {
        for (int i = 0; i < particles.length; i++) {
            particles[i] = new Particle();
        }
    }

    class Particle {
        private PVector pos;
        private PVector vel;
        private PVector acc;

        private PVector oldpos;

        private float maxSpeed;

        private int col;


        public Particle() {
            pos = new PVector(random(width-1), random(height-1));
            vel = new PVector(0, 0);
            acc = new PVector(0, 0);

            oldpos = new PVector(pos.x, pos.y);

            maxSpeed = 4;

            col = color(random(255), 255, 255);
        }

        public void follow(PVector[] vectors) {
            int x = floor(pos.x/scl);
            int y = floor(pos.y/scl);

            int index = x + y * cols;
            if(index >= vectors.length){
                index %= vectors.length;
            }

            applyForce(vectors[index]);
        }

        public void applyForce(PVector force) {
            acc.add(force);
        }

        public void display() {
            stroke(col, 5);
            strokeWeight(1);
            line(pos.x, pos.y, oldpos.x, oldpos.y);

            updateOldPos();
        }

        public void updateColor(int c) {
            col = c;
        }

        public void update() {
            vel.add(acc);
            vel.limit(maxSpeed);
            pos.add(vel);
            acc.mult(0);
        }

        public void updateOldPos() {
            oldpos.x = pos.x;
            oldpos.y = pos.y;
        }

        public void edges() {
            if (pos.x > width-1) {
                pos.x = 1;
                updateOldPos();
            }

            if (pos.x < 1) {
                pos.x = width-1;
                updateOldPos();
            }

            if (pos.y > height-1) {
                pos.y = 1;
                updateOldPos();
            }

            if (pos.y < 1) {
                pos.y = height-1;
                updateOldPos();
            }
        }
    }
}

File: src/FlowField/FlowField.js
const scl = 20;

const fieldStrength = 0.3;

var zoffInc = 0.01;

var angleMult = 2;

var zoff = 0;
var noiseInc = 0.1;

var rows, cols;

var flowfield = [];
var particles = [];

var h = 0;
var colUp = true;

const maxSpeed = 4;

const num = 10000;

var mm = 0;

function setup() {
    createCanvas(windowWidth, windowHeight);
    cols = width / scl;
    rows = height / scl;
    background(0);
    colorMode(HSB, 255);
    createParticles();
}

function draw() {
    background(0, 0, 0, 15);

    var yoff = 0;
    for (var y = 0; y < rows; y++) {
        var xoff = 0;
        for (var x = 0; x < cols; x++) {
            var index = x + y * cols;
            var a = noise(xoff, yoff, zoff) * TWO_PI * angleMult;
            xoff += noiseInc;

            var v = p5.Vector.fromAngle(0);
            v.rotate(a);
            v.setMag(fieldStrength);

            flowfield[index] = v;
        }
        yoff += noiseInc;
    }
    zoff += zoffInc;

    for (var particle of particles) {
        particle.updateColor(color(h, 255, 255));
        particle.edges();
        particle.follow(flowfield);
        particle.update();
        particle.display();
    }

    if (colUp) {
        h++;
    } else {
        h--;
    }

    if (h > 255) {
        colUp = false;
        h = 255;
    } else if (h < 0) {
        colUp = true;
        h = 0;
    }
}

function mouseClicked() {
    var temp = num / 100;
    for (var i = mm; i < num; i += temp) {
        particles[i].updateloc(mouseX, mouseY);
    }
    mm = (mm + 1) % temp;
}

function refresh() {
    background(0);
    createParticles();
}

function createParticles() {
    for (var i = 0; i < 10000; i++) {
        particles[i] = new Particle();
    }
}

function Particle() {
    this.pos = createVector(random(width - 1), random(height - 1));
    this.vel = createVector(0, 0);
    this.acc = createVector(0, 0);
    this.oldpos = createVector(this.pos.x, this.pos.y);
    this.col = color(random(255), 255, 255);

    this.updateloc = function (x, y) {
        this.pos.x = x;
        this.pos.y = y;
        this.updateOldPos()
        var temp = random(0, TWO_PI);
        this.vel.x = maxSpeed * cos(temp);
        this.vel.y = maxSpeed * sin(temp);
        this.acc.mult(0);
    }

    this.follow = function (vets) {
        var x = floor(this.pos.x / scl);
        var y = floor(this.pos.y / scl);

        var index = x + y * cols;
        if (index >= vets.length) {
            index %= vets.length;
        }

        this.applyForce(vets[index]);
    }

    this.applyForce = function (force) {
        this.acc.add(force);
    }

    this.display = function () {
        stroke(this.col, 5);
        strokeWeight(1);
        line(this.pos.x, this.pos.y, this.oldpos.x, this.oldpos.y);
        this.updateOldPos();
    }

    this.updateColor = function (c) {
        this.col = c;
    }

    this.update = function () {
        this.vel.add(this.acc);
        this.vel.limit(maxSpeed);
        this.pos.add(this.vel);
        this.acc.mult(0);
    }

    this.updateOldPos = function () {
        this.oldpos.x = this.pos.x;
        this.oldpos.y = this.pos.y;
    }

    this.edges = function () {
        if (this.pos.x > width - 1) {
            this.pos.x = 1;
            this.updateOldPos();
        }

        if (this.pos.x < 1) {
            this.pos.x = width - 1;
            this.updateOldPos();
        }

        if (this.pos.y > height - 1) {
            this.pos.y = 1;
            this.updateOldPos();
        }

        if (this.pos.y < 1) {
            this.pos.y = height - 1;
            this.updateOldPos();
        }
    }
}